actors/builtin/miner/miner_actor.go

```diff
@@ -160,53 +173,164 @@ func (a Actor) ChangePeerID(rt Runtime, params *ChangePeerIDParams) *adt.EmptyVa
 // WindowedPoSt //
 //////////////////
 
+// Information submitted by a miner to provide a Window PoSt.
+type SubmitWindowedPoStParams struct {
+	// The partition indices being proven.
+	// Partitions are counted across all deadlines, such that all partition indices in the second deadline are greater
+	// than the partition numbers in the first deadlines.
+	Partitions []uint64
+	// Parallel array of proofs corresponding to the partitions.
+	Proofs []abi.PoStProof
+	// Sectors skipped while proving that weren't already declared faulty
+	Skipped abi.BitField
+}
+
 // Invoked by miner's worker address to submit their fallback post
-func (a Actor) SubmitWindowedPoSt(rt Runtime, params *abi.OnChainWindowPoStVerifyInfo) *adt.EmptyValue {
+func (a Actor) SubmitWindowedPoSt(rt Runtime, params *SubmitWindowedPoStParams) *adt.EmptyValue {
+	if uint64(len(params.Partitions)) > WPoStMessagePartitionsMax {
+		rt.Abortf(exitcode.ErrIllegalArgument, "too many partitions %d, max %d", len(params.Partitions), WPoStMessagePartitionsMax)
+	}
+	if len(params.Partitions) != len(params.Proofs) {
+		rt.Abortf(exitcode.ErrIllegalArgument, "proof count %d must match partition count %", len(params.Proofs), len(params.Partitions))
+	}
+
+	currEpoch := rt.CurrEpoch()
+	store := adt.AsStore(rt)
 	var st State
+	var detectedFaultSectors []*SectorOnChainInfo
+	var penalty abi.TokenAmount
+	var recoveredSectors []*SectorOnChainInfo
 	rt.State().Transaction(&st, func() interface{} {
 		rt.ValidateImmediateCallerIs(st.Info.Worker)
 
-		if rt.CurrEpoch() > st.PoStState.ProvingPeriodStart+power.WindowedPostChallengeDuration {
-			rt.Abortf(exitcode.ErrIllegalState, "PoSt submission too late")
+		// Every epoch is during some deadline's challenge window.
```

Comment:
I would strongly recommend having the user submit the partition they are targeting. Its likely just one extra byte in the parameters, and will make diagnosing issues *much* easier

Reply:
The _partitions_ are explicitly submitted, but assuming you meant deadline: 👍  I'll do that.